### PR TITLE
[RateLimiter] Added reserve() to LimiterInterface and rename Limiter to RateLimiter

### PR DESCRIPTION
**Pull request by wouterj**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

While Javier wrote documentation for this new component, we found a couple of confusing elements that might need some tweaks:

* The `Limiter` class (previously called `LimiterFactory`) has imho a bit strange name, as it's not a limiter and it doesn't implement `LimiterInterface`. It can only new limiters. I believe `LimiterFactory` - like `LockFactory` - would be the most clear, but as that was rejected before, here is another proposal using `RateLimiter`.
* `reserve()` was now only part of the token bucket implementation. That made it a bit less useful. I think I've found a way to also allow reserving future hits in the fixed window implementation, so I've moved it to the `LimiterInterface`.

### CODE REVIEW

File: src/Symfony/Component/RateLimiter/CompoundLimiter.php [hunk 1]
```diff
@@ -31,6 +33,11 @@ public function __construct(array $limiters)
         $this->limiters = $limiters;
     }
 
+    public function reserve(int $tokens = 1, ?float $maxTime = null): Reservation
+    {
+        // TODO: Implement reserve() method.
+    }
```
Comment (wouterj) [inline src/Symfony/Component/RateLimiter/CompoundLimiter.php]: I didn't really know how a compound limiter should act on reserving tokens. Should it reserve a token for each limiter? That can lead to reserving tokens at all different timeslots, which timeslot should this limiter then return?

Or should we add a `ReserveNotSupportedException` that we throw in such cases? (which might also be needed for other implementations, like https://github.com/symfony/symfony/pull/38565 or leaky bucket)

**Comment by Nyholm** [comment]
👍  for `RateLimiter`

**Comment by nicolas-grekas** [comment]
(rebase needed)

File: src/Symfony/Component/RateLimiter/CompoundLimiter.php (hunk 1 above)
Comment (fabpot) [reply src/Symfony/Component/RateLimiter/CompoundLimiter.php]: At first, I would indeed throw an exception. That way, if we find a good way later on, we will be able to support it "easily" without any BC breaks.

**Review by fabpot — APPROVED** [review]
Ready IMHO when the TODO is resolved

File: src/Symfony/Component/RateLimiter/Exception/MaxWaitDurationExceededException.php [hunk 2]
```diff
@@ -11,11 +11,27 @@
 
 namespace Symfony\Component\RateLimiter\Exception;
 
+use Symfony\Component\RateLimiter\Limit;
+use Throwable;
```
Comment (derrabus) [inline src/Symfony/Component/RateLimiter/Exception/MaxWaitDurationExceededException.php]: We usually don't import classes from the global namespace.

File: src/Symfony/Component/RateLimiter/FixedWindowLimiter.php [hunk 3]
```diff
@@ -33,49 +34,72 @@ final class FixedWindowLimiter implements LimiterInterface
 
     public function __construct(string $id, int $limit, \DateInterval $interval, StorageInterface $storage, ?LockInterface $lock = null)
     {
+        if ($limit < 1) {
+            throw new \InvalidArgumentException(sprintf('Cannot set the limit of "%s" to 0, as that would never accept any hit.', TokenBucketLimiter::class));
```
Comment (derrabus) [inline src/Symfony/Component/RateLimiter/FixedWindowLimiter.php]:
```suggestion
            throw new \InvalidArgumentException(sprintf('Cannot set the limit of "%s" to 0, as that would never accept any hit.', __CLASS__));
```

Wrong class: This is the constructor of `FixedWindowLimiter`.

**Review by derrabus — CHANGES_REQUESTED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @wouterj.

**Pull request merged**